### PR TITLE
Remove python-idaklu

### LINKED ISSUE #4223: remove python-idaklu
### Description

The python-idaklu solver is a legacy solver that uses idaklu but calls back into python for the functions. With the recent improvements to idaklu this can now be removed.

### Motivation

reduce the number of solvers we need to maintain

### Possible Implementation

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by pipliggins**
# Description

Removes the legacy python-idaklu solver which calls IDAKLU but returns back to python for certain functions. With recent improvements to IDAKLU this can now be removed.

Reduces the number of solvers we need to maintain.

Fixes #4223 

## Type of change

- [x] Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] Optimization (back-end change that speeds up the code)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by kratman** [comment]
@pipliggins What still needs to be done for this PR?

**Comment by pipliggins** [comment]
@kratman I think it's pretty much ready to go. The only query (which I realise now I hadn't actually voiced anywhere, whoops) is what to do about some indirect changes flagged by codecov which leave several lines of code untested after removing the solver.
It's not clear to me that those lines are specific to the old IDAKLU solver, in fact I'm pretty sure at least some of them shouldn't be, but I've found it hard to work out why they aren't covered by other tests, or how to reach those lines by adding new ones. (@martinjrobins any thoughts as well?)

**Comment by kratman** [comment]
> The only query (which I realise now I hadn't actually voiced anywhere, whoops) is what to do about some indirect changes flagged by codecov which leave several lines of code untested after removing the solver. It's not clear to me that those lines are specific to the old IDAKLU solver, in fact I'm pretty sure at least some of them shouldn't be, but I've found it hard to work out why they aren't covered by other tests, or how to reach those lines by adding new ones

I think a lot of the test coverage is actually indirect. Some tests are only run with specific parameter sets or certain modes might only be run in a test for something unrelated. Ultimately as large functions get broken up into smaller once, this should get better.

Personally a loss of a little coverage does not bother me, but I do wonder if some of the lost coverage could be from logic that was catching rare edge cases.

File: src/pybamm/solvers/idaklu_solver.py
Comment (kratman) [inline src/pybamm/solvers/idaklu_solver.py:756]: If this is not needed anymore then it should be removed
```suggestion
```

File: src/pybamm/solvers/idaklu_solver.py
Comment (kratman) [inline src/pybamm/solvers/idaklu_solver.py:771]:
```suggestion
            raise pybamm.SolverError("Unsupported IDAKLU solver configuration.")
```

File: src/pybamm/solvers/idaklu_solver.py
Comment (kratman) [reply src/pybamm/solvers/idaklu_solver.py:300]: Alternatively the `self._options["jax_evaluator"] != "iree"` check could be added here:
```
        if model.convert_to_format == "jax":
            mass_matrix = model.mass_matrix.entries.toarray()
```

**Review by kratman — COMMENTED** [review]
Just minor suggestions. The coverage changes seem fairly small, but @martinjrobins should take a look as well

File: src/pybamm/models/base_model.py
Comment (agriyakhetarpal) [reply src/pybamm/models/base_model.py:42]: I think this was a missing docstring that was never added before this PR

Comment (kratman) [reply src/pybamm/models/base_model.py:42]: Yeah updating the docs is always good. Just checking if there was added functionality somewhere in the PR to go along with it

**Review by martinjrobins — CHANGES_REQUESTED** [review]
Thanks @pipliggins, looking forward to no longer supporting this :) The codecov errors reminded me: now the python idaklu solver is removed we no longer use the python evaluator in sensitivitiy calculations, this is why some of these covereage errors. I would suggest to raise an error after line 1471 in base_solver.py (if model.calculate_sensitivities is true) saying that sensitivity calculation for the python evaluator is no longer supported, since we don't want to add tests for this. We can then remove this if statement from coverage with `# pragma: no cover`

File: src/pybamm/models/base_model.py
Comment (pipliggins) [reply src/pybamm/models/base_model.py:42]: No added functionality, just noticed the description was missing as I worked my way though relevant sections

**Review by martinjrobins — APPROVED** [review]
thanks @pipliggins, this looks good to me. Once the conflicts are resolved it can be merged.

**Comment by pipliggins** [comment]
@martinjrobins @kratman Conflicts have been resolved, this is ready to merge. Thanks!

**Review by kratman — APPROVED** [review]
Thanks looks good

**Pull request merged**
